In [2]:
import pandas as pd
import numpy as np

In [3]:
df=pd.read_csv("https://catalog.ourworldindata.org/garden/covid/latest/compact/compact.csv")

In [29]:
df

,country,date,total_cases,new_cases,new_cases_smoothed,total_cases_per_million,new_cases_per_million,new_cases_smoothed_per_million,total_deaths,new_deaths,...,population,population_density,median_age,life_expectancy,gdp_per_capita,extreme_poverty,diabetes_prevalence,handwashing_facilities,hospital_beds_per_thousand,human_development_index
0,Afghanistan,2020-01-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
1,Afghanistan,2020-01-02,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
2,Afghanistan,2020-01-03,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
3,Afghanistan,2020-01-04,0.0,0.0,NaN,0.000,0.0,NaN,0.0,0.0,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
4,Afghanistan,2020-01-05,0.0,0.0,NaN,0.000,0.0,NaN,0.0,0.0,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
589774,Zimbabwe,2026-05-06,266436.0,0.0,0.0,16580.682,0.0,0.0,5740.0,0.0,...,16069061.0,41.538223,17.683001,62.360100,5215.252930,49.219894,1.5,42.509991,1.95,NaN
589775,Zimbabwe,2026-05-07,266436.0,0.0,0.0,16580.682,0.0,0.0,5740.0,0.0,...,16069061.0,41.538223,17.683001,62.360100,5215.252930,49.219894,1.5,42.509991,1.95,NaN
589776,Zimbabwe,2026-05-08,266436.0,0.0,0.0,16580.682,0.0,0.0,5740.0,0.0,...,16069061.0,41.538223,17.683001,62.360100,5215.252930,49.219894,1.5,42.509991,1.95,NaN
589777,Zimbabwe,2026-05-09,266436.0,0.0,0.0,16580.682,0.0,0.0,5740.0,0.0,...,16069061.0,41.538223,17.683001,62.360100,5215.252930,49.219894,1.5,42.509991,1.95,NaN


# **Basic Exploratory Data Analysis**

## <u>*Dataset Info*</u> 

In [70]:
df.shape

(589779, 61)

In [23]:
df.head(5)

,country,date,total_cases,new_cases,new_cases_smoothed,total_cases_per_million,new_cases_per_million,new_cases_smoothed_per_million,total_deaths,new_deaths,...,population,population_density,median_age,life_expectancy,gdp_per_capita,extreme_poverty,diabetes_prevalence,handwashing_facilities,hospital_beds_per_thousand,human_development_index
0,Afghanistan,2020-01-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
1,Afghanistan,2020-01-02,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
2,Afghanistan,2020-01-03,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
3,Afghanistan,2020-01-04,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
4,Afghanistan,2020-01-05,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN


In [99]:
df.info()
df["date"]=pd.to_datetime(df["date"]) 
# converting date from object to datetime

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 589779 entries, 0 to 589778
Data columns (total 61 columns):
 #   Column                                      Non-Null Count   Dtype         
---  ------                                      --------------   -----         
 0   country                                     589779 non-null  object        
 1   date                                        589779 non-null  datetime64[ns]
 2   total_cases                                 577431 non-null  float64       
 3   new_cases                                   573183 non-null  float64       
 4   new_cases_smoothed                          571974 non-null  float64       
 5   total_cases_per_million                     577431 non-null  float64       
 6   new_cases_per_million                       573183 non-null  float64       
 7   new_cases_smoothed_per_million              571974 non-null  float64       
 8   total_deaths                                577431 non-null  float64      

In [100]:
df.isnull().sum().sort_values(ascending=False)

human_development_index                    589779
weekly_icu_admissions                      578786
weekly_icu_admissions_per_million          578786
excess_mortality_cumulative_per_million    576007
excess_mortality                           575973
                                            ...  
total_cases_per_million                     12348
total_deaths                                12348
total_deaths_per_million                    12348
date                                            0
country                                         0
Length: 61, dtype: int64

***Many column have missing value more than 50% of total data***

In [101]:
size=np.round(df.memory_usage(deep=True).sum()/1024**2,2)
print(f"DataFrame(df) size : {size} MB")

DataFrame(df) size : 353.61 MB


In [102]:
df.duplicated().sum()

np.int64(0)

- *Observation*<br>
    - The shape of DataFrame is (589779, 61).<br>
    - Many column have missing value more than 50% of total data.<br>
    - Size of DataFrame :382.3 MB
***

***
## <u>Columns Analysis</u>

- **Objective**<br>
    - *Checking each Column thorougly and getting important information out from it.*<br>
    - *Get important Pattern out from different columns.*<br>

### Country AND Date Analysis

In [103]:
countries=df["country"].unique()
no_of_unique_country=df["country"].nunique()
no_of_unique_country

262

In [104]:
print(f"Maximum days data: {df["country"].value_counts().max()}")
print(f"Minimum days data: {df["country"].value_counts().min()}")
print("for each country")

Maximum days data: 2322
Minimum days data: 1
for each country


In [105]:
country_count=df["country"].value_counts()
Countries=country_count[country_count==2322].index
df_complete=df[df["country"].isin(Countries)]
df_complete["country"].value_counts()
Countries.shape

(181,)

In [106]:
missing_count=df.groupby("country")["total_cases"].apply(lambda x : x.isnull().sum())
missing_count.sort_values(ascending=False).head(5)

country
Hong Kong           1672
Taiwan              1386
Northern Ireland    1385
England             1359
Scotland            1327
Name: total_cases, dtype: int64

-  *Observation*
    - The maximum number of missing values in `total_cases` is observed for Hong Kong (1672 missing records).
    - Several countries contain substantial missing data and may require filtering before comparative analysis.
    - 181 countries out of total 262 countries(Involving Olympic Event And World Data) have complete records covering all 2322 days.

In [107]:
last_date_of_record=df["date"].max()   
first_date_of_record=df["date"].min()

In [108]:
country_date_range = (
    df.groupby("country")["date"]
      .agg(
          start_date="min",
          end_date="max",
          count="count"
      )
      .reset_index()
)

country_date_range

,country,start_date,end_date,count
0,Afghanistan,2020-01-01,2026-05-10,2322
1,Africa,2020-01-04,2026-05-10,2319
2,Albania,2020-01-01,2026-05-10,2322
3,Algeria,2020-01-01,2026-05-10,2322
4,American Samoa,2020-01-04,2026-05-10,2319
...,...,...,...,...
257,World excl. China and South Korea,2020-01-04,2026-05-10,2319
258,"World excl. China, South Korea, Japan and Sing...",2020-01-04,2026-05-10,2319
259,Yemen,2020-01-01,2026-05-10,2322
260,Zambia,2020-01-01,2026-05-10,2322


In [109]:
(last_date_of_record-first_date_of_record).days

2321

- *Observation*
    - Most countries have records from 2020-01-01 to 2026-05-10.
    - Some countries/entities started reporting later (e.g., Africa starts on 2020-01-04).
    - Countries with fewer records may indicate missing or incomplete reporting.

### Case Analysis

In [142]:
total_cases_max = df.groupby("country")["total_cases"].max()

total_cases_max.sort_values(ascending=False).head(15)
# total_cases_max["India"]
Cases_per_country=(df.groupby("country").agg(
    Total_Cases = ("total_cases","max"),
    Max_New_Cases= ("new_cases","max")
).reset_index())
peak_dates=df.loc[
    df.groupby("country")["new_cases"].idxmax().dropna(), 
    ["country","date"]
]
# used dropna bcz peak_dates was getting some NAN values
Cases_per_country=Cases_per_country.merge(peak_dates,on="country")
# Cases_per_country[Cases_per_country["country"]=="India"]
Cases_per_country.sort_values(by="Max_New_Cases",ascending=False).head(10)

,country,Total_Cases,Max_New_Cases,date
242,World,779131743.0,8401906.0,2022-01-30
243,World excl. China,679749982.0,8401713.0,2022-01-30
244,World excl. China and South Korea,645178109.0,8384638.0,2022-01-30
245,"World excl. China, South Korea, Japan and Sing...",608368382.0,8295098.0,2022-01-30
12,Asia,302181607.0,7213802.0,2022-12-23
96,High-income countries,458049345.0,7197333.0,2022-01-30
234,Upper-middle-income countries,239829083.0,7027032.0,2022-12-23
45,China,99381761.0,6966046.0,2022-12-23
70,Europe,254960190.0,6386449.0,2022-01-30
71,European Union (27),186985473.0,5896096.0,2022-01-30


In [149]:
Cases_per_country[Cases_per_country["country"].isin(["China","India"])]

,country,Total_Cases,Max_New_Cases,date
45,China,99381761.0,6966046.0,2022-12-23
100,India,45056221.0,414188.0,2021-05-07


In [156]:
country_cases=Cases_per_country[~Cases_per_country["country"].str.contains("World|excl.|Europe|Asia|North|South|income|Upper|European")]
country_cases.sort_values(by="Total_Cases",ascending=False)


,country,Total_Cases,Max_New_Cases,date
232,United States,103436829.0,1265520.0,2022-01-12
45,China,99381761.0,6966046.0,2022-12-23
100,India,45056221.0,414188.0,2021-05-07
76,France,39060446.0,2417043.0,2022-01-23
82,Germany,38437953.0,1588891.0,2022-03-27
...,...,...,...,...
159,Niue,1092.0,117.0,2023-01-02
220,Tokelau,80.0,39.0,2023-07-15
238,Vatican,26.0,7.0,2020-10-13
175,Pitcairn,4.0,4.0,2022-07-20


In [164]:
print(df["new_cases_smoothed"].max())
print(df["new_cases"].max())
df[["date","new_cases","new_cases_smoothed"]].sort_values(by="new_cases",ascending=False).head(8)

6402033.0
8401906.0


,date,new_cases,new_cases_smoothed
574294,2022-01-30,8401906.0,3346287.8
576613,2022-01-30,8401713.0,3346053.2
578932,2022-01-30,8384638.0,3331154.5
574287,2022-01-23,8346273.0,3362941.2
576606,2022-01-23,8345987.0,3362752.2
578925,2022-01-23,8338479.0,3356269.5
581251,2022-01-30,8295098.0,3254791.2
581244,2022-01-23,8286572.0,3311735.8


- *Obsevation*
    - Highest case in a day reported by any Individual country is China and which is On 2022-12-2023
    - Total cases in India was 45056221 and maximum new_cases found in a day in India was 414188 On 2021-05-07
    - In dataset(df) there is a data of different continents which will always be more then individual countries so for different country comparison we need to filter out countries.
    - United States had highest number of cases followed by China and India.(It shows that USA,China and India were most affected)
    - Smoothed features shows the seven days cummulative avg of all the new cases such that it become smoothed new cases.Because Covid do not exploded suddenly so we can get better trend analysis with this smoothed data